src/range_proof/mod.rs

```diff
@@ -184,48 +285,76 @@ impl RangeProof {
             .zip(util::exp_iter(y.invert()))
             .zip(concat_z_and_2)
-            .map(|((s_i_inv, exp_y_inv), z_and_2)| z + exp_y_inv * (zz * z_and_2 - b * s_i_inv));
+            .map(|((s_i_inv, exp_y_inv), z_and_2)| {
+                z + exp_y_inv * (zz * z_and_2 - b * s_i_inv)
+            });
+
+        let value_commitment_scalars = util::exp_iter(z)
+            .take(m)
+            .map(|z_exp| c * zz * z_exp);
 
-        let value_commitment_scalars = util::exp_iter(z).take(m).map(|z_exp| c * zz * z_exp);
         let basepoint_scalar = w * (self.t_x - a * b) + c * (delta(n, m, &y, &z) - self.t_x);
 
-        let mega_check = ristretto::vartime::multiscalar_mul(
-            iter::once(Scalar::one())
+        // Challenge value for combining the complete range proof statement with other range proof statements.
+        let batch_challenge = Scalar::random(rng);
+
+        Verification {
+            pedersen_base_scalars: (
+                batch_challenge*basepoint_scalar, 
+                batch_challenge*(-self.e_blinding - c * self.t_x_blinding)
+            ),
+            g_scalars: g.map(|s| batch_challenge*s ).collect(),
+            h_scalars: h.map(|s| batch_challenge*s ).collect(),
+            dynamic_base_scalars: iter::once(Scalar::one())
                 .chain(iter::once(x))
                 .chain(value_commitment_scalars)
                 .chain(iter::once(c * x))
                 .chain(iter::once(c * x * x))
-                .chain(iter::once(-self.e_blinding - c * self.t_x_blinding))
-                .chain(iter::once(basepoint_scalar))
-                .chain(g)
-                .chain(h)
                 .chain(x_sq.iter().cloned())
-                .chain(x_inv_sq.iter().cloned()),
-            iter::once(&self.A)
+                .chain(x_inv_sq.iter().cloned())
+                .map(|s| batch_challenge*s )
+                .collect(),
+            dynamic_bases: iter::once(&self.A)
                 .chain(iter::once(&self.S))
-                .chain(value_commitments.iter())
+                .chain(value_commitments.as_ref().iter())
                 .chain(iter::once(&self.T_1))
                 .chain(iter::once(&self.T_2))
-                .chain(iter::once(&gens.pedersen_generators.B_blinding))
-                .chain(iter::once(&gens.pedersen_generators.B))
-                .chain(gens.G.iter())
-                .chain(gens.H.iter())
                 .chain(self.ipp_proof.L_vec.iter())
-                .chain(self.ipp_proof.R_vec.iter()),
-        );
-
-        if mega_check.is_identity() {
-            Ok(())
-        } else {
-            Err(())
+                .chain(self.ipp_proof.R_vec.iter())
+                .cloned()
+                .collect(),
         }
     }
 }
 
+/// Represents a deferred computation to verify a single rangeproof.
+/// Multiple instances can be verified more efficient as a batch using
+/// `RangeProof::verify_batch` function.
+struct Verification {
```

Comment:
In the actual batchverify code above, each of these fields are used only once in the inner loop accumulating the batched verification scalars.

So, instead of having a seperate `Verification` struct that contains owned copies of the verification scalars for a given `RangeProof`, why not just have functions on the `RangeProof` struct that return the iterators for each of the fields of what is currently the `Verification` struct?